apps/meeting/src/providers/DataMessagesProvider/index.tsx

```diff
@@ -0,0 +1,109 @@
+import { RosterType, useAudioVideo, useMeetingManager, useRosterState } from 'amazon-chime-sdk-component-library-react';
+import { DataMessage } from 'amazon-chime-sdk-js';
+import React, { useEffect, useReducer, createContext, useContext, FC, useCallback, useRef } from 'react';
+import { DataMessagesActionType, initialState, ChatDataMessage, reducer } from './state';
+
+interface DataMessagesStateContextType {
+  sendMessage: (message: string) => void;
+  messages: ChatDataMessage[];
+}
+
+const DataMessagesStateContext = createContext<DataMessagesStateContextType | undefined>(undefined);
+
+export const DataMessagesProvider: FC = ({ children }) => {
+  const meetingManager = useMeetingManager();
+  const audioVideo = useAudioVideo();
+  const [state, dispatch] = useReducer(reducer, initialState);
+  const { roster } = useRosterState();
+  const rosterRef = useRef<RosterType>({});
+
+  useEffect(() => {
+    if (!audioVideo) {
+      return;
+    }
+    audioVideo.realtimeSubscribeToReceiveDataMessage('ChimeComponentLibraryDataMessage', handler);
+    return () => {
+      audioVideo.realtimeUnsubscribeFromReceiveDataMessage('ChimeComponentLibraryDataMessage');
+    };
+  }, [audioVideo]);
+
+  useEffect(() => {
+    if (roster) {
+      rosterRef.current = { ...rosterRef.current, ...roster };
+    }
+  }, [roster]);
+
+  const handler = useCallback(
+    (dataMessage: DataMessage) => {
+      if (!dataMessage.throttled) {
+        const isSelf =
+          dataMessage.senderAttendeeId === meetingManager.meetingSession?.configuration.credentials?.attendeeId;
+        if (isSelf) {
+          dispatch({
+            type: DataMessagesActionType.ADD,
+            payload: {
+              message: new TextDecoder().decode(dataMessage.data),
+              senderAttendeeId: dataMessage.senderAttendeeId,
+              timestamp: dataMessage.timestampMs,
+              senderName: rosterRef.current[dataMessage.senderAttendeeId]?.name!,
+              isSelf: true,
+            },
+          });
+        } else {
+          const data = dataMessage.json();
+          dispatch({
+            type: DataMessagesActionType.ADD,
+            payload: {
+              message: data.message,
+              senderAttendeeId: dataMessage.senderAttendeeId,
+              timestamp: dataMessage.timestampMs,
+              senderName: rosterRef.current[dataMessage.senderAttendeeId]?.name!,
+              isSelf: false,
+            },
+          });
+        }
```

Comment:
I think we can merge these two conditions.
```js
dispatch({
  type: DataMessagesActionType.ADD,
  payload: {
    message: isSelf? new TextDecoder().decode(dataMessage.data): data.message, // <====
    senderAttendeeId: dataMessage.senderAttendeeId,
    timestamp: dataMessage.timestampMs,
    senderName: rosterRef.current[dataMessage.senderAttendeeId]?.name!,
    isSelf: isSelf, // <====
  },
});
```

Reply:
Actually, in `else` here, I have to do `dataMessage.json()` to get the `data` which is not the case for the "self" path. The self path received DataMessage's message is a Uint8Array and not stringified JSON hence handled in separate cases.

Reply:
Oh that's right!